### PR TITLE
wip commit to suppress mau alerting for small instances

### PR DESCRIPTION
**Pull request by neilisfragile**
### Pull Request Checklist

* [ ] Pull request is based on the develop branch
* [ ] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [ ] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

### CODE REVIEW

File: synapse/config/server.py [hunk 1]
```diff
@@ -695,9 +696,15 @@ def generate_config_section(
...
         #
+        # 'mau_suppress_alerting' is a means of limiting client side alerting
+        # should the mau limit be reached. This is useful for small instances
+        # where the admin has 5 mau seats (say) for 5 specific people and no
+        # interest increasing the mau limit further. Defaults to False.
+        #
         #limit_usage_by_mau: False
         #max_mau_value: 50
         #mau_trial_days: 2
+        #mau_suppress_alerting: True
```
Comment (anoadragon453) [inline synapse/config/server.py]: It might make more intuitive sense to have the option be to alert, with it defaulting to True?

Comment (babolivier) [reply synapse/config/server.py]: Agreed, iirc lately we've mainly been trying to add options to enable stuff (which can default to `True`) rather than ones to disable stuff to avoid confusion.

Comment (neilisfragile) [reply synapse/config/server.py]: My thinking was that almost all cases alerting will be the default so the option made sense to be in the negative. Sounds like it is confusing, so I have inverted.

File: synapse/server_notices/resource_limits_server_notices.py [hunk 2]
```diff
@@ -91,6 +91,15 @@ def maybe_send_server_notice_to_user(self, user_id):
         currently_blocked, ref_events = yield self._is_room_currently_blocked(room_id)
 
         try:
+            if not self._config.mau_limit_alerting:
```
Comment (richvdh) [inline synapse/server_notices/resource_limits_server_notices.py]: the big stack of if statements in this function is getting really hard to follow, and I'm worried that the duplicated code will get out of sync.

I think it would be clearer to make the `check_auth_blocking` call (ie, everything between lines 103 and 112 as it currently stands) conditional on `mau_limit_alerting` (leaving `is_auth_blocking = False` if `mau_limit_alerting` is false).

Comment (neilisfragile) [reply synapse/server_notices/resource_limits_server_notices.py]: @richvdh I've had a go a refactoring the whole area, I _think_ it is now easier to read, albeit still inherently fiddly.

File: synapse/server_notices/resource_limits_server_notices.py [hunk 3]
```diff
@@ -91,47 +91,77 @@ def maybe_send_server_notice_to_user(self, user_id):
...
+    def _remove_limit_block_notification(self, user_id, ref_events):
+        """Utility method to remove limit block notifications from the server
+        notices room.
+
+        Args:
+            user_id (str): user to notify
+            ref_events ([]): The list of pinned events that are unrelated to
+            limit blocking and need to be preserved.
+        """
+        content = {"pinned": ref_events}
```
Comment (richvdh) [inline synapse/server_notices/resource_limits_server_notices.py]: this seems to be an existing problem (so not necessarily for fixing as part of this PR), but why do we preserve the other events in this case and not in the `_apply_limit_block_notification` case?

File: synapse/server_notices/resource_limits_server_notices.py [hunk 4]
```diff
@@ -91,47 +91,77 @@ def maybe_send_server_notice_to_user(self, user_id):
...
             logger.error("Error sending resource limits server notice: %s", e)
 
+    @defer.inlineCallbacks
+    def _remove_limit_block_notification(self, user_id, ref_events):
+        """Utility method to remove limit block notifications from the server
+        notices room.
+
+        Args:
+            user_id (str): user to notify
+            ref_events ([]): The list of pinned events that are unrelated to
```
Comment (richvdh) [inline synapse/server_notices/resource_limits_server_notices.py]:
```suggestion
            ref_events (list[str]): The event_ids of pinned events that are unrelated to
```

File: synapse/server_notices/resource_limits_server_notices.py [hunk 5]
```diff
@@ -91,47 +91,77 @@ def maybe_send_server_notice_to_user(self, user_id):
...
-                event_content = e.msg
-                event_limit_type = e.limit_type
+            if not self._config.mau_limit_alerting:
+                # Alerting disabled, reset room if necessary and return
+                if currently_blocked:
+                    self._remove_limit_block_notification(ref_events, user_id)
+                return
+
+            is_auth_blocking = False
+            if self._config.mau_limit_alerting:
```
Comment (richvdh) [inline synapse/server_notices/resource_limits_server_notices.py]: this condition is redundant, it will always be `True`

File: synapse/server_notices/resource_limits_server_notices.py (hunk 2 above)
Comment (richvdh) [reply synapse/server_notices/resource_limits_server_notices.py]: I guess you didn't agree with my suggestion? Fair enough. It's certainly clearer now.

**Comment by richvdh** [comment]
Unrelated to the change at hand, but isn't this thing massively racy? if two calls `maybe_send_server_notice_to_user` happen at once, we might get several copies of the server notice?

File: synapse/server_notices/resource_limits_server_notices.py (hunk 3 above)
Comment (neilisfragile) [reply synapse/server_notices/resource_limits_server_notices.py]: From memory the idea was to shut everything down and so all tags should be discarded, but in retrospect this seems unhelpful.

File: synapse/server_notices/resource_limits_server_notices.py (hunk 2 above)
Comment (neilisfragile) [reply synapse/server_notices/resource_limits_server_notices.py]: For reference https://gist.github.com/richvdh/092934e994a2443bd184a01357c98677 Agreed that proposal is neater, but less explicit about the blocking behaviour, so will not adopt.

File: synapse/server_notices/resource_limits_server_notices.py (hunk 3 above)
Comment (neilisfragile) [reply synapse/server_notices/resource_limits_server_notices.py]: fixed

**Comment by neilisfragile** [comment]
> Unrelated to the change at hand, but isn't this thing massively racy? if two calls `maybe_send_server_notice_to_user` happen at once, we might get several copies of the server notice?

I can't think when multiple overlapping calls would occur, but if they did I think it would result in a more confusing dag but continue to work as expected.

File: synapse/server_notices/resource_limits_server_notices.py (hunk 3 above)
Comment (neilisfragile) [reply synapse/server_notices/resource_limits_server_notices.py]: unfixed, will address in a separate PR.

**Comment by richvdh** [comment]
> I can't think when multiple overlapping calls would occur

the user syncs from two devices at once?

> but if they did I think it would result in a more confusing dag but continue to work as expected.

wouldn't there be two copies of the server notice?

File: synapse/server_notices/resource_limits_server_notices.py [hunk 6]
```diff
@@ -91,24 +92,28 @@ def maybe_send_server_notice_to_user(self, user_id):
...
+                # to other users if they were to arrive.
+                yield self._auth.check_auth_blocking()
+            except ResourceLimitError as e:
+                is_auth_blocking = True
+                event_body = e.msg
+                event_limit_type = e.limit_type
+
+            if (
+                not self._config.mau_limit_alerting
+                and event_limit_type is LimitBlockingTypes.MONTHLY_ACTIVE_USER
```
Comment (richvdh) [inline synapse/server_notices/resource_limits_server_notices.py]: use `==` not `is` except when checking for None

```suggestion
                and event_limit_type == LimitBlockingTypes.MONTHLY_ACTIVE_USER
```

File: synapse/server_notices/resource_limits_server_notices.py [hunk 7]
```diff
@@ -91,24 +92,28 @@ def maybe_send_server_notice_to_user(self, user_id):
         currently_blocked, ref_events = yield self._is_room_currently_blocked(room_id)
 
         try:
-            if not self._config.mau_limit_alerting:
-                # Alerting disabled, reset room if necessary and return
+            is_auth_blocking = False
+            event_limit_type = ""
```
Comment (richvdh) [inline synapse/server_notices/resource_limits_server_notices.py]: `None` would be a more conventional sentinel here

```suggestion
            event_limit_type = None
```

File: synapse/server_notices/resource_limits_server_notices.py [hunk 8]
```diff
@@ -91,24 +92,28 @@ def maybe_send_server_notice_to_user(self, user_id):
         currently_blocked, ref_events = yield self._is_room_currently_blocked(room_id)
 
         try:
-            if not self._config.mau_limit_alerting:
-                # Alerting disabled, reset room if necessary and return
+            is_auth_blocking = False
```
Comment (richvdh) [inline synapse/server_notices/resource_limits_server_notices.py]: this feels pretty redundant - indeed, you don't bother to check it at line 110. Let's just get rid of it and use `event_limit_type` instead.

File: synapse/server_notices/resource_limits_server_notices.py (hunk 7 above)
Comment (richvdh) [reply synapse/server_notices/resource_limits_server_notices.py]: shouldn't it be called `limit_type` rather than `event_limit_type`?
